embassy-rp/src/adc.rs

```diff
@@ -42,24 +46,55 @@ impl<'p> Pin<'p> {
             w.set_pue(pull == Pull::Up);
             w.set_pde(pull == Pull::Down);
         });
-        Self { pin: pin.map_into() }
+        Self(Source::Pin(pin.map_into()))
+    }
+
+    pub fn new_sensor(s: impl Peripheral<P = ADC_TEMP_SENSOR> + 'p) -> Self {
+        let r = pac::ADC;
+        r.cs().write_set(|w| w.set_ts_en(true));
+        Self(Source::TempSensor(s.into_ref()))
     }
 
     fn channel(&self) -> u8 {
-        // this requires adc pins to be sequential and matching the adc channels,
-        // which is the case for rp2040
-        self.pin._pin() - 26
+        match &self.0 {
+            // this requires adc pins to be sequential and matching the adc channels,
+            // which is the case for rp2040
+            Source::Pin(p) => p._pin() - 26,
+            Source::TempSensor(_) => 4,
+        }
     }
 }
 
-impl<'d> Drop for Pin<'d> {
+impl<'p> Drop for Source<'p> {
     fn drop(&mut self) {
-        self.pin.pad_ctrl().modify(|w| {
-            w.set_ie(true);
-            w.set_od(false);
-            w.set_pue(false);
-            w.set_pde(true);
-        });
+        match self {
+            Source::Pin(p) => {
+                p.pad_ctrl().modify(|w| {
+                    w.set_ie(true);
+                    w.set_od(false);
+                    w.set_pue(false);
+                    w.set_pde(true);
+                });
+            }
+            Source::TempSensor(_) => {
+                pac::ADC.cs().write_clear(|w| w.set_ts_en(true));
+            }
+        }
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Debug, Default)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[repr(transparent)]
+pub struct Sample(u16);
+
+impl Sample {
+    pub fn good(&self) -> bool {
+        self.0 < 0x8000
```

Comment:
datasheet says the "error" bit is bit 12, not bit 15

Reply:
hah, I had an old version of the datasheet :joy: 
old:
![screenshot-2023-07-31_02-51-32](https://github.com/embassy-rs/embassy/assets/1247578/ff4b6586-f8f3-40b1-b3ad-345323f12a5d)
new:
![screenshot-2023-07-31_02-53-33](https://github.com/embassy-rs/embassy/assets/1247578/7a3fb909-8df9-41ab-b6a9-7ae4b6df169e)


the bit is indeed 15 (it's the same in both versions)
![screenshot-2023-07-31_02-52-13](https://github.com/embassy-rs/embassy/assets/1247578/78e1fde0-2b35-4f88-bfd6-ddd435508c9d)